src/Linker.DefineFunction.tt

```diff
@@ -0,0 +1,310 @@
+﻿<#@ template debug="false" hostspecific="false" language="C#" #>
+<#@ assembly name="System.Core" #>
+<#@ import namespace="System.Collections.Generic" #>
+<#@ import namespace="System.Globalization" #>
+<#@ import namespace="System.Text" #>
+<#@ import namespace="System.Linq" #>
+<#@ output extension=".cs" encoding="us-ascii" #>
+<#
+// Note: We specify ASCII as output encoding above to prevent different UTF-8 BOM behavior with VS and dotnet-t4.
+#>
+// <auto-generated>
+// This file is autogenerated from a T4 text template (Linker.DefineFunction.tt).
+// Do not modify it directly.
+// To generate the file outside from Visual Studio, install and run dotnet-t4:
+//
+//     dotnet tool install --global dotnet-t4 --version 2.3.0
+//     t4 src/Linker.DefineFunction.tt -o src/Linker.DefineFunction.cs
+//
+// </auto-generated>
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace Wasmtime
+{
+    public partial class Linker
+    {
+<#
+// Generate overloads for different combinations of function parameter count, result count,
+// and hasCaller states.
+// Currently, we generate them for a maximum of 12 parameters and 4 result values, which means there
+// will be 2 * 13 * 5 = 130 overloads of DefineFunction accepting an Action/Func.
+// Note that more than 200 overloads can cause the C# compiler to no longer resolve the overloads
+// correctly (tested with Visual Studio 17.4.0 Preview 2.1).
+// For delegate types not covered by these overloads, we have a method accepting a Delegate which
+// will be called using reflection.
+const int maxParameterCount = 12;
+const int maxReturnCount = 4;
+
+for (int hasCaller = 0; hasCaller < 2; hasCaller++)
```

Comment:
Nit: perhaps this more clearly be expressed as `foreach (var hasCaller in new[] { true, false })`?